interpreter.py

```diff
@@ -51,59 +73,111 @@ def exit_not_implemented(self, name):
         elif opcode == LLVMAdd:
             x, y = self._get_args(args, state)
-            return x + y
+            assert isinstance(x, NumericType) and isinstance(y, NumericType)
+            return Integer(x.value + y.value)
+        elif opcode == LLVMFAdd:
+            x, y = self._get_args(args, state)
+            assert isinstance(x, NumericType) and isinstance(y, NumericType)
+            return Float(x.value + y.value)
         elif opcode == LLVMMul:
             x, y = self._get_args(args, state)
-            return x * y
+            assert isinstance(x, NumericType) and isinstance(y, NumericType)
+            return Integer(x.value * y.value)
+        elif opcode == LLVMSub:
+            x, y = self._get_args(args, state)
+            assert isinstance(x, NumericType) and isinstance(y, NumericType)
+            return Integer(x.value - y.value)
         elif opcode == LLVMCall:
-            string_format_ref = LLVMGetOperand(args[0], 0)
-            string_format = lookup_var(state, self.global_state, string_format_ref)
-            fn_name = rffi.charp2str(LLVMGetValueName(args[-1]))
-            if fn_name == "printf":
-                # XXX assumes all printf arguments are integers
-                printf_args = []
-                for arg in args[1:-1]:
-                    printf_args.append(lookup_var(state, self.global_state, arg))
-                print string_format % tuple(printf_args)
+            if rffi.cast(rffi.INT, args[-1]) in self.functions.keys():
+                print "Found function", rffi.charp2str(LLVMGetValueName(args[-1]))
+                for index in range(0, LLVMCountParams(args[-1])):
+                    param = LLVMGetParam(args[-1], index)
+                    self.global_state.set_variable(rffi.cast(rffi.INT, param),\
+                                                   lookup_var(state, self.global_state, args[index]))
+                interp_fun = Interpreter(self.functions, self.global_state)
+                return interp_fun.run(args[-1])
             else:
-                exit_not_implemented(fn_name)
+                print rffi.charp2str(LLVMGetValueName(args[-1]))
+                string_format_ref = LLVMGetOperand(args[0], 0)
+                str_var = lookup_var(state, self.global_state, string_format_ref)
+                assert isinstance(str_var, String)
+                string_format = str_var.value
+                fn_name = rffi.charp2str(LLVMGetValueName(args[-1]))
+                if fn_name == "printf":
+                    printf_args = []
+                    for i in range(1, len(args) - 1):
+                        arg = args[i]
+                        var = lookup_var(state, self.global_state, arg)
+                        printf_args.append(var)
+                    self.puts(string_format, printf_args)
+                elif fn_name == "puts":
+                    self.puts(string_format)
+                else:
+                    self.exit_not_implemented(fn_name)
```

Comment:
This all looks good, but I wonder if we should take the large call logic out if the dispatch loop and into a method for readability.